backend/downsample_test.py

```diff
@@ -0,0 +1,303 @@
+        assert test_result == [records[2], records[5]]
+
+        max_records = 100
+        test_result = max_min_downsample(records, True, max_records)
+        assert test_result == records
+
+        max_records = 100
+        test_result = max_min_downsample([], True, max_records)
+        assert test_result == []
+
+    def test_min_downsample(self, records):
+        """Tests on max_min_downsample method, only work on minimum downsample"""
+        max_records = 0
+        test_result = max_min_downsample(records, False, max_records)
+        assert test_result == []
+
+        max_records = 1
+        test_result = max_min_downsample(records, False, max_records)
+        assert test_result == [records[8]]
+
+        max_records = 2
+        test_result = max_min_downsample(records, False, max_records)
+        assert test_result == [records[0], records[8]]
+
+        max_records = 100
+        test_result = max_min_downsample(records, False, max_records)
+        assert test_result == records
+
+        max_records = 100
+        test_result = max_min_downsample([], False, max_records)
+        assert test_result == []
+
+    def test_average_downsample(self, records):
+        """Tests on average_downsample method"""
+        max_records = 0
+        test_result = average_downsample(records, max_records)
+        assert test_result == []
+
+        max_records = 1
+        test_result = average_downsample(records, max_records)
+        expected_average = [
+            [
+                sum([record[0] for record in records]) / len(records),
+                sum([record[1] for record in records]) / len(records),
+                records[0][2]
+            ]
+        ]
+        assert test_result == expected_average
+
+        max_records = 2
+        test_result = average_downsample(records, max_records)
+
+        # Calculates average over first and second halves.
+        expected_average = [
+            [
+                sum([record[0] for record in records[:len(records) // 2]]
+                    ) / (len(records) // 2),
+                sum([record[1] for record in records[:len(records) // 2]]
+                    ) / (len(records) // 2),
+                records[0][2]
+            ],
+            [
+                sum([record[0] for record in records[len(records) // 2:]]
+                    ) / (len(records) // 2),
+                sum([record[1] for record in records[len(records) // 2:]]
+                    ) / (len(records) // 2),
+                records[0][2]
+            ],
+        ]
+        assert test_result == expected_average
+
+        max_records = 100
+        test_result = average_downsample(records, max_records)
+        expected_average = records
+        assert test_result == expected_average
+
+        max_records = 100
+        test_result = average_downsample([], max_records)
+        assert test_result == []
+
+    def assert_records_in_each_second(self, filename, max_records_per_second):
```

Comment:
this method is testing downsample, so better rename is to something like test_downsample_records_sampling_per_second

Reply:
This method won't work without a temporary file setup, so I consider this method as a helper for assertion.
Also renaming to `test_something` would make pytest treat this method as an individual test, which would break for the above reason.